GPU examples

Giesch opened the pull request:
This adds / will add some more SDL GPU examples based on the official ones in C.

Notes:
- the shader json metadata is from shadercross
  I've been using a justfile setup for managing that in my own project, we could do something similar here,
  or I could remove the serde dependencies and hardcode the resource counts.
  The serde dependency is also the reason for the new examples directory ([based on axum's examples setup](https://github.com/tokio-rs/axum)).
- my editor/LSP setup locally was sad without a rustfmt.toml specifying the edition

Remaining todos:
- [x] use some more rust-friendly patterns (ie avoid unnecessary `SDL_free`)
- [x] re-add the clear screen example
- [x] re-add the basic triangle example
- [x] find a way to share the common module
- [x] ~~maybe add a shadercross setup for compiling example shaders~~
  (edit: I'm going to think about how to approach this, and maybe do a separate PR)
- [x] maybe switch examples to using app_impl
- [x] remove build-from-source-static 
- [x] resolve bugs on macos

Giesch committed:
clear screen example

Giesch committed:
match hyphen convention

Giesch committed:
add basic triangle example

Giesch committed:
remove unused fixed timestep from triangle example

Giesch committed:
lint for clear screen example

Giesch committed:
Merge remote-tracking branch 'origin/main' into gpu-examples

Giesch committed:
Merge remote-tracking branch 'origin/main' into gpu-examples

Giesch committed:
cargo update

Giesch committed:
broken with linking error

Giesch committed:
update pull-sprite-batch to match sdl3-experiment

Giesch committed:
remove inner examples setup without dependencies

Giesch committed:
fix sdl src version

Lokathor commented:
if there should be a rustfmt file just add a rustfmt file

Giesch committed:
Merge remote-tracking branch 'upstream/main' into gpu-examples

Giesch committed:
add basic triangle example

* use workspace root for content dir

Giesch commented:
Sorry; I can see how that'd come off as catty. I meant it as an explanation for a kind of pointless change. I don't think the file's actually required by rustfmt itself, only by my editor extension.

Lokathor commented:
I'm not the lead but the crates are using 2021 edition, so there should probably be a rustfmt that configures for that.

Giesch committed:
use std::fs for loading shader code

Giesch committed:
specify correct edition in rustfmt.toml

Giesch committed:
cargo fmt

maia-s commented:
Looks good so far! Thank you for doing this.

I don't think the examples should require the `build-from-source-static` feature so people can run them with an already installed SDL library, but you can leave taking that out for last if it makes developing easier for you.

maia-s commented:
The examples fail here (macos) with ``newLibraryWithSource:8657: failed assertion `source must not be nil.'``. I can investigate later.

Giesch committed:
avoid unused import warning on windows

Giesch committed:
use app_impl in basic triangle

Giesch committed:
use app_impl in pull sprite batch

Giesch committed:
fix --release assertion failure in basic triangle

* the color_target_descriptions array needs its own declaration
  without this, the slice pointer gets invalidated in release mode
  which leads to the texture format being some arbitrary value

Giesch committed:
use explicit const for title strings

* now I'm paranoid about as_ptr

Giesch committed:
avoid same inline as_ptr in pull sprite batch

Giesch committed:
use explicit common namespace, set framerate hint

Giesch committed:
add clear screen example

Giesch committed:
remove build-from-source-static

Giesch committed:
handle shader file path correctly based on format

Giesch commented:
> The examples fail here (macos) with `` newLibraryWithSource:8657: failed assertion `source must not be nil.' ``. I can investigate later.

I hadn't been handling non-vulkan shader paths at all. The most recent commit fixes this error for the clear screen & basic triangle examples.

But, now the sprite batch example fails less directly (on macos). It builds and runs, but seems scaled (way) too large.

Giesch committed:
fix deinit_gpu_window copy/paste

Giesch committed:
add repr(C, packed) to GPUSprite struct

* this fixes the macos scaling bug

Giesch committed:
add a random seed matching the upstream example

Giesch commented:
The macos bug was due to alignment. There may be a more rust-friendly way to handle the GPUSprite struct, but this seems like the way that matches the upstream example.

maia-s commented:
I see why `repr(C)` is required, but does `packed` do anything there? I'd expect it to be tightly packed anyway since all the fields are `f32`s

Giesch committed:
remove unnecessary 'packed'

Giesch commented:
Nope, it was unnecessary. I should spend some time with the docs.

Giesch commented:
Are there other changes I should make here? Or are these kinds of examples more useful on the higher-level `sdl3` crate?

maia-s commented:
Oh, sorry, I didn't realize you were done with the changes you wanted to make. I've been busy moving to France and dealing with things related to that, but I'll look at this soon.

maia-s commented:
Sorry for taking so long to get to this! I've looked it over and it looks good for the most part, so I'm going to merge it now. I had to rebuild the shaders for the pull-sprite example to work on macos and there's some minor reorganization I want to do, but I think you've been waiting long enough so we can deal with that after the merge 😅 Thank you for your work!

I'm going to move this repo to Codeberg in the near future, so if you have any further pull requests you want to do, I'd appreciate it if you make them there instead when that's set up. I'll keep this github repo as a mirror for the forseeable future.
